sql/core/src/main/scala/org/apache/spark/sql/sources/CombineSmallFile.scala

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources
+
+import org.apache.hadoop.fs.{FileStatus, FileSystem, Path}
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.SQLContext
+
+object CombineSmallFile {
+  def combineWithFiles[T](rdd: RDD[T], sqlContext: SQLContext, inputFiles: Array[FileStatus])
+      : RDD[T] = {
+    if (sqlContext.conf.combineSmallFile) {
+      val totalLen = inputFiles.map { file =>
+        if (file.isDir) 0L else file.getLen
+      }.sum
+      val numPartitions = (totalLen / sqlContext.conf.splitSize + 1).toInt
+      rdd.coalesce(numPartitions)
```

Comment:
I think this is a very hack way to solve this problem. As we can not tell how the the data source to be split, even for Hadoop, the split size just a hint, use that for computing the partition number probably too risky for a generic data process framework.

And the `RDD.coalesce` actually will combine the splits in a arbitrary way, it's probably causes the data skew, as we most likely combine the large partitions into a a single task.

IMO, I'd like to deep investigate how Hive to combine the small partitions, by using the `CombineHiveInputFormat` or `HiveInputFormat`, which seems has a strategy to select the partitions according to both input format, and also keep the balance.